### ISSUE TITLE
Question: Named pipes: await JsonSerializer.DeserializeAsync<T> blocks, when reading from Named pipe

### ISSUE THREAD

**Issue by Legends** [dotnet/runtime#81840]
I have created a named pipe server (console app, .NET 7):

```
using var server = new NamedPipeServerStream(
                                                            "hostpipe",
                                                            PipeDirection.InOut,
                                                            NamedPipeServerStream.MaxAllowedServerInstances,
                                                            PipeTransmissionMode.Message,
                                                            PipeOptions.Asynchronous);
```

that spawns a child process which creates a `NamedPipeClientStream? :
`using var client = new NamedPipeClientStream(".", pipeName, PipeDirection.InOut, PipeOptions.Asynchronous);`

**Problem:**
For example I want to message an object from the server to the client like:

Server writes:
public static async Task WriteMessageAsyncAsObject<T>(PipeStream pipe, T obj)
{
    await JsonSerializer.SerializeAsync<T>(pipe, obj); // works!
 }

and the client should read:
 public static async Task<T?> ReadMessageAsyncAsObject<T>(PipeStream pipe)
        {
            // var bt = await ReadMessageAsync(pipe);
            // var obj = JsonSerializer.Deserialize<T>(bt);
            // return await Task.FromResult(obj);
            return await JsonSerializer.DeserializeAsync<T>(pipe);  // doesn't work, the program stops here, it is blocked!
        }

But if I replace `JsonSerializer.DeserializeAsync<T>` on the clientside with the following:
```
public static async Task<T?> ReadMessageAsyncAsObject<T>(PipeStream pipe)
{
    var bt = await ReadMessageAsync(pipe);  // reads the stream result asynchronously into a MemoryStream, which returns a byte[]
    var obj = JsonSerializer.Deserialize<T>(bt);  // works now, but synchronous!
    return await Task.FromResult(obj);
    //return await JsonSerializer.DeserializeAsync<T>(pipe);  
}
```

**Why is `JsonSerializer.DeserializeAsync<T>` blocking on the clientside?**

**Comment by wfurt** [comment]
I don't think this is IO problem since `ReadMessageAsync` works fine according to description. 
This is similar to https://github.com/dotnet/runtime/issues/73097 and we end up fixing it in HTTP.

**Comment by Jozkee** [comment]
I wasn't able to repro this. Perhaps, there's something special about the type you are serializing.

Program 1:
```cs
using System.Diagnostics;
using System.IO.Pipes;
using System.Text.Json;

namespace ConsoleApp4
{
    internal class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Hello, World!");

            using var server = new NamedPipeServerStream(
                                    "hostpipe",
                                    PipeDirection.InOut,
                                    NamedPipeServerStream.MaxAllowedServerInstances,
                                    PipeTransmissionMode.Message,
                                    PipeOptions.Asynchronous);

            using Process p = Process.Start(@"C:\consoleapps\ConsoleApp5\ConsoleApp5\bin\Debug\net7.0\ConsoleApp5.exe");
            server.WaitForConnection();

            var person = new Person
            {
                Age = 42,
                Name = "Foo",
            };
            await WriteMessageAsyncAsObject(server, person);
        }

        public static async Task WriteMessageAsyncAsObject<T>(PipeStream pipe, T obj)
        {
            await JsonSerializer.SerializeAsync<T>(pipe, obj); // works!
        }
    }

    internal class Person
    {
        public int Age { get; set; }
        public string Name { get; set; }
    }
}
```

Program2:
```cs
using System.Diagnostics;
using System.IO.Pipes;
using System.Text.Json;

namespace ConsoleApp5
{
    internal class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Hello, World!");

            using var client = new NamedPipeClientStream(".", "hostpipe", PipeDirection.InOut, PipeOptions.Asynchronous);
            client.Connect();

            var r = await ReadMessageAsyncAsObject<Person>(client);

            Console.WriteLine(r.Age);
            Console.WriteLine(r.Name);
        }

        public static async Task<T?> ReadMessageAsyncAsObject<T>(PipeStream pipe)
        {
            // var bt = await ReadMessageAsync(pipe);
            // var obj = JsonSerializer.Deserialize<T>(bt);
            // return await Task.FromResult(obj);
            return await JsonSerializer.DeserializeAsync<T>(pipe);  // doesn't work, the program stops here, it is blocked!
        }
    }

    internal class Person
    {
        public int Age { get; set; }
        public string Name { get; set; }
    }
}
```

Output:
```
Hello, World!
Hello, World!
42
Foo
```

**Comment by wfurt** [comment]
It may be the size @Jozkee. If you are on it, I would try something that would perhaps fill the pipe buffers  - maybe 100k ish.

**Comment by Jozkee** [comment]
@wfurt gave it a quick try with `new NamedPipeServerStream(..., inBufferSize: 1, outBufferSize: 1)`, no error. Also tried serializing a `new Person { Age = 42, Name = new string('a', 100_000) }`, still no error :/. 

Would be nice to hear back from OP and see if he/she can provide a repro.

**Comment by Legends** [comment]
I have created a repo [here](https://github.com/Legends/Blocking-Issue).

**Comment by eiriktsarpalis** [comment]
I can reproduce the issue, the application hangs when the serializer tries to read the underlying stream for a second time here:

https://github.com/dotnet/runtime/blob/f4ad730ac2c3f06fe68a3041c21590ef9de1b8c2/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs#L51-L57

The underlying implementation in Windows is this method:

https://github.com/dotnet/runtime/blob/f4ad730ac2c3f06fe68a3041c21590ef9de1b8c2/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs#L314-L366

Without looking much into the code, I suspect there might be an issue with the `ReadWriteValueTaskSource` implementation that prevents the STJ callback to be scheduled appropriately.

**Comment by stephentoub** [comment]
@eiriktsarpalis, I'm not sure why this would be considered a System.IO.Pipes problem.  The JsonSerializer.DeserializeAsync enters this loop:
https://github.com/dotnet/runtime/blob/ac7afb9ccb88b895eeb3264e38ab22d0c5d726ec/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs#L49-L67
It reads the 50 bytes the server sent and checks the while loop condition.  The `fillBuffer` argument is defaulting to `true`, so since it only read 50 bytes and was given a 16384 byte buffer, it loops around again to perform another read, which hangs, because there's nothing more to read.

**Comment by eiriktsarpalis** [comment]
Ah yes, the async deserializer generally assumes that the stream only contains a single JSON document and will always try to fill its internal buffer up to capacity or EOF _before_ it starts deserializing. This is due to how async converters are designed and was meant to improve deserialization performance.

Even though the internal buffer size can be controlled via the `JsonSerializerOptions.DefaultBufferSize` property, setting this to a smaller number would still make deserialization susceptible to the same hangs under [certain circumstances](https://github.com/dotnet/runtime/blob/ac7afb9ccb88b895eeb3264e38ab22d0c5d726ec/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs#L114-L130). Deserialization could also fail if it detects [trailing data after the first complete JSON document](https://github.com/dotnet/runtime/blob/463954dbc728470fb21a097188fc3dae91f6cbe4/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs#L1731-L1734).

TL;DR the `DeserializeAsync` methods have not been designed to read from streams that represent channels and which could contain multiple values. The `DeserializeAsyncEnumerable` method does support this (and does not impose the same "fillBuffer" semantics) although it imposes certain constraints (the stream must be one JSON array) and it would make your application much less straightforward to implement. Alternatively you might want to consider filling a buffer manually and passing that to one of the sync `Deserialize` methods, although some work

In the future it might be possible to support your scenario in the regular `DeserializeAsync` methods by adding:

1. A `JsonSerializerOptions` flag that disables "fillBuffer" semantics.
2. A `JsonSerializerOptions` flag that disables failures if the stream contains trailing data.

Related to #36750, #33030.

**Comment by RobinHSanner** [comment]
I was having the same issue and resolved it by liberally sprinkling .ConfigureAwait(false) sauce on everything in the call chain.
